### PR TITLE
crimson/osd/osd_operations/client_request_common: `PeeringState::needs_recovery()` may fail if the object is under backfill

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/66216
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by Matan-B — CHANGES_REQUESTED** [review]
This function is hard to read, what do you think about:
```
  bool needs_recovery_or_backfill = false;

  if (pg->is_unreadable_object(soid)) {
    auto &peering_state = pg->get_peering_state();
    auto &missing_loc = peering_state.get_missing_loc();
    ceph_assert(missing_loc.needs_recovery(soid, &ver));
    needs_recovery_or_backfill = true;
  }
  
  if (pg->is_degraded_or_backfilling_object(soid)) {
    needs_recovery_or_backfill = true;
  }
  
  if (recover_or_backfill == false) {
    logger().debug("{} reqid {} nothing to recover {}",
                   __func__, reqid, soid);
    return seastar::now();
  }
  
  ...
```

---
For context, 7743d33136e7e2431cca1c3929adba6cd386e773 recently added the call to `needs_recovery()` in order to obtain the correct `ver`.

**Comment by xxhdx1985126** [comment]
Done, please take a look, again, thanks.

File: src/crimson/osd/osd_operations/client_request_common.cc [hunk 1]
```diff
@@ -25,14 +25,24 @@ CommonClientRequest::do_recover_missing(
...
+    auto &missing_loc = peering_state.get_missing_loc();
+    ceph_assert(missing_loc.needs_recovery(soid, &ver));
+    needs_recovery_or_backfill = true;
+  }
+
+  if (pg->is_degraded_or_backfilling_object(soid)) {
+    needs_recovery_or_backfill = true;
+  }
+
+  if (needs_recovery_or_backfill == false) {
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/client_request_common.cc:41]:
```suggestion
  if (!needs_recovery_or_backfill) {
```

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request_common.cc:41]: Ah, yes.

**Review by Matan-B — APPROVED** [review]
LGTM

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-57691/

**Comment by xxhdx1985126** [comment]
> osd-backfill test keeps failing with this change:
> 
> * https://pulpito.ceph.com/matan-2024-06-10_10:58:00-crimson-rados-wip-matanb-crimson-testing-57691-distro-crimson-smithi/
> * https://pulpito.ceph.com/matan-2024-06-10_12:01:35-crimson-rados-wip-matanb-crimson-testing-57691-distro-crimson-smithi/7747924/
> 
> I'm not sure that the actual change here is causing the issue. It looks to me that we just expose a different bug here. Moreover, I have seen the _exact_ assertion fails in the classical OSD as well (not yet resolved).
> 
> See: https://tracker.ceph.com/issues/50608
> 
> @xuxuehan, what do you think?
> 
> Fix attempt: #57269

Seems that the OSDs' log is not on the qa-proxy site, do you have a copy?

**Comment by Matan-B** [comment]
> Seems that the OSDs' log is not on the qa-proxy site, do you have a copy?

https://qa-proxy.ceph.com/teuthology/matan-2024-06-10_10:58:00-crimson-rados-wip-matanb-crimson-testing-57691-distro-crimson-smithi/7747884/remote/smithi138/log/

**Comment by Matan-B** [comment]
Ah, got it:
osd.1 is crashed with 
```
DEBUG 2024-06-10 12:21:49,771 [shard 0:main] osd - handle_push: MOSDPGPush(1.1 30/29 {PushOp(1:8d34ce13:::benchmark_data_smithi173_33788_object2355:head, version: 27'627, data_included: [0~4096], data_size: 4096, omap_header_size: 0, omap_entries_size: 0, attrset_size: 2, recovery_info: ObjectRecoveryInfo(1:8d34ce13:::benchmark_data_smithi173_33788_object2355:head@0'0, size: 4096, copy_subset: [(0, 4096)], clone_subset: {}, snapset: 0=[]:{}, object_exist: false), after_progress: ObjectRecoveryProgress(!first, data_recovered_to: 4096, data_complete: true, omap_recovered_to: , omap_complete: true, error: false), before_progress: ObjectRecoveryProgress(first, data_recovered_to: 0, data_complete: false, omap_recovered_to: , omap_complete: false, error: false))}) v4
DEBUG 2024-06-10 12:21:49,771 [shard 0:main] osd - _handle_push
DEBUG 2024-06-10 12:21:49,771 [shard 0:main] osd - submit_push_data
...
ceph-osd: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4192-ga62e0e55/rpm/el9/BUILD/ceph-19.0.0-4192-ga62e0e55/src/crimson/osd/pg_recovery.cc:371: RecoveryBackend::interruptible_future<> PGRecovery::on_local_recover(const hobject_t&, const ObjectRecoveryInfo&, bool, ceph::os::Transaction&): Assertion `pg->is_primary()' failed.
```

osd.2 starts this recovery with empty version:
```
DEBUG 2024-06-10 12:21:49,655 [shard 0:main] osd - do_recover_missing reqid client.4517.0:2356 object 1:8d34ce13:::benchmark_data_smithi173_33788_object2355:head version 0'0, starting recovery
```

Which causes the following condition in `PGRecovery::on_local_recover` to be true:
```
      log.get_missing().get_items().find(recovery_info.soid)->second.need > recovery_info.version) {
    assert(pg->is_primary());
```

**Comment by Matan-B** [comment]
@xxhdx1985126, I have pushed a commit which should help. Let me know what do you think, feel free to squash or remove.

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126, I have pushed a commit which should help. Let me know what do you think, feel free to squash or remove.
> 
> Results look good: backfill: https://pulpito.ceph.com/matan-2024-06-17_12:20:20-crimson-rados-wip-matanb-crimson-testing-57367-v2-distro-crimson-smithi/
> 
> Suite: https://pulpito.ceph.com/matan-2024-06-17_12:19:19-crimson-rados-wip-matanb-crimson-testing-57367-v2-distro-crimson-smithi/

Great！Just squashed, thanks:-)

**Review by Matan-B — APPROVED** [review]

**Review by athanatos — APPROVED** [review]

**Pull request merged**